### PR TITLE
Build metrics with opts

### PR DESCRIPTION
**Pull request by overvenus**
This PR consists of:
 1. ~~Drop support for untype.~~ (Moved to #57) 
 2. Refactor vec, build metrics with opts.

This is the beginning of #55 .

### CODE REVIEW

**Comment by siddontang** [comment]
Please do one thing in one PR.

**Comment by siddontang** [comment]
Please merge master first.

**Comment by siddontang** [comment]
I think we can simplify the code like:

+ Add a variable names in Opts
+ Generate the Desc from Opts (maybe use move like `fn desc(self) -> Desc`)
+ Pass Opts, no need to pass Desc

File: src/desc.rs [hunk 1]
```diff
@@ -137,3 +137,9 @@ impl Desc {
         Ok(desc)
     }
 }
+
+/// `Describer` describes the immutable meta-data of a Metric.
+pub trait Describer {
+    /// `desc` returns a `Desc`.
```
Comment (siddontang) [inline src/desc.rs]: describe

**Comment by siddontang** [comment]
We can pass Opt to Value::new directly, no need Desc.

**Comment by siddontang** [comment]
LGTM

**Comment by siddontang** [comment]
PTAL @disksing

File: src/metrics.rs [hunk 2]
```diff
@@ -74,6 +75,14 @@ pub struct Opts {
     /// that label most likely should not be a label at all (but part of the
     /// metric name).
     pub const_labels: HashMap<String, String>,
+
+    /// variable_labels contains names of labels for which the metric maintains
+    /// variable values. Metrics with the same fully-qualified name must have
+    /// the same label names in their variable_labels.
+    ///
+    /// Note that variable_labels is used in `MetricVec`. To create a single
+    /// metric must leaves it empty.
```
Comment (disksing) [inline src/metrics.rs]: must **leave** it empty.

**Comment by siddontang** [comment]
PTAL @disksing

**Comment by disksing** [comment]
LGTM

**Pull request merged**